fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -0,0 +1,356 @@
+     * <b>Note:</b> {@link #loadPartitioningMetadata(Set)} should be called
+     * before the first call to this method.
+     * If no partition currently exists, a new one is created.
+     *
+     * @param grouping document grouping key
+     * @param timestamp document timestamp
+     * @return assigned partition metadata
+     */
+    public LucenePartitionInfoProto.LucenePartitionInfo assignWritePartition(Tuple grouping, long timestamp) {
+        List<LucenePartitionInfoProto.LucenePartitionInfo> partitions = partitionMetaInfoCache.get(grouping);
+        if (partitions == null) {
+            // if loadPartitioningMetadata(...) was called earlier, then this should never happen
+            throw new RecordCoreException("Internal Error");
+        }
+
+        if (partitions.isEmpty()) {
+            // very first partition
+            partitions.add(LucenePartitionInfoProto.LucenePartitionInfo.newBuilder()
+                    .setId(0)
+                    .setCount(0)
+                    .setFrom(timestamp)
+                    .setTo(timestamp)
+                    .build());
+            return partitions.get(0);
+        }
+
+        return selectWritePartition(partitions, timestamp);
+        // TO DO: check and split the assigned partition if it's too big
+    }
+
+    /**
+     * add a timestamp to the metadata of a given partition and save to db.
+     * The <code>count</code> will be incremented, and the <code>from</code> or <code>to</code> timestamps will
+     * be adjusted if applicable.
+     *
+     * @param grouping  grouping key
+     * @param partition partition metadata
+     * @param timestamp document timestamp
+     */
+    public void addToAndSavePartitionMetadata(final Tuple grouping, final LucenePartitionInfoProto.LucenePartitionInfo partition, final Long timestamp) {
+        LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partition.toBuilder();
+        builder.setCount(partition.getCount() + 1);
+        if (timestamp < partition.getFrom()) {
+            builder.setFrom(timestamp);
+        }
+        if (timestamp > partition.getTo()) {
+            builder.setTo(timestamp);
+
+        }
+        savePartitionMetadataAndUpdateCache(grouping, builder);
+    }
+
+    /**
+     * remove a document from a partition metadata and save to db.
+     * Note that only the document count is changed (decremented). <code>from</code> and <code>to</code> are unchanged.
+     *
+     * @param grouping grouping key
+     * @param partition partition metadata object
+     */
+    public void removeFromAndSavePartitionMetadata(final Tuple grouping, final LucenePartitionInfoProto.LucenePartitionInfo partition) {
+        LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partition.toBuilder();
+        // note that the to/from of the partition do not get updated, since that would require us to know what the next potential boundary
+        // value(s) are.
```

Comment:
This would only produce a change in the metadata if the removed document happens to have the max or the min timestamp of the partition. The range of the partition remains correct in all cases, which is why I think it may be ok to let it stay like this (at least for now).
Eventually, when we split partitions, the metadata would need to be updated, but that would be simpler.

Reply:
That makes sense.
It basically means that you may have partitions with ranges larger than the min/max of the actual data, but it will never be less, so you can still guarantee a strict ordering of the partitions.
And the splitting will need to update, but it will have to select the x oldest or newest documents anyways, at which point it can correctly update the metadata for the partition.